lib/kubernetes-deploy/render_task.rb

```diff
@@ -0,0 +1,101 @@
+# frozen_string_literal: true
+require 'tempfile'
+
+require 'kubernetes-deploy/renderer'
+require 'kubernetes-deploy/template_discovery'
+
+module KubernetesDeploy
+  class RenderTask
+    def initialize(logger:, current_sha:, template_dir:, bindings:)
+      @logger = logger
+      @template_dir = template_dir
+      @renderer = KubernetesDeploy::Renderer.new(
+        current_sha: current_sha,
+        bindings: bindings,
+        template_dir: @template_dir,
+        logger: @logger,
+      )
+    end
+
+    def run(*args)
+      run!(*args)
+      true
+    rescue KubernetesDeploy::FatalDeploymentError
+      false
+    end
+
+    def run!(stream, only_filenames = [])
+      @logger.reset
+      @logger.phase_heading("Initializing render task")
+
+      filenames = if only_filenames.empty?
+        TemplateDiscovery.new(@template_dir).templates
+      else
+        only_filenames
+      end
+
+      validate_configuration(filenames)
+      @logger.phase_heading("Rendering template(s)")
+
+      filenames.each { |filename| render_filename(filename, stream) }
+
+      @logger.summary.add_action("Successfully rendered #{filenames.size} template(s)")
+      @logger.print_summary(:success)
+    rescue KubernetesDeploy::FatalDeploymentError
+      @logger.print_summary(:failure)
+      raise
+    end
+
+    private
+
+    def render_filename(filename, stream)
+      file_content = File.read(File.join(@template_dir, filename))
+      rendered_content = @renderer.render_template(filename, file_content)
+      YAML.load_stream(rendered_content, "<rendered> #{filename}") do |doc|
+        stream.puts YAML.dump(doc)
+      end
+      @logger.info("Rendered #{File.basename(filename)}")
```

Comment:
Not sure how I missed this in the last review, but I think we change this to `Rendering  ...` and put it at the top. Having an empty phase 2 here isn't awesome:

```
Begin test: test_render_invalid_yaml
***************************************************************************

[INFO][2018-11-16 10:47:50 -0800]	
[INFO][2018-11-16 10:47:50 -0800]	---------------------------------Phase 1: Initializing render task----------------------------------
[INFO][2018-11-16 10:47:50 -0800]	Validating configuration
[INFO][2018-11-16 10:47:50 -0800]	
[INFO][2018-11-16 10:47:50 -0800]	-----------------------------------Phase 2: Rendering template(s)-----------------------------------
[INFO][2018-11-16 10:47:50 -0800]	
[INFO][2018-11-16 10:47:50 -0800]	------------------------------------------Result: FAILURE-------------------------------------------
[FATAL][2018-11-16 10:47:50 -0800]	Invalid template: yaml-error.yml
[FATAL][2018-11-16 10:47:50 -0800]	Error message: (<rendered> yaml-error.yml): mapping values are not allowed in this context at line 9 column 21
```

Reply:
Added that statement, output looks like this now:

```
[INFO][2018-12-11 11:36:42 -0500]
[INFO][2018-12-11 11:36:42 -0500]	---------------------------------Phase 1: Initializing render task----------------------------------
[INFO][2018-12-11 11:36:42 -0500]	Validating configuration
[INFO][2018-12-11 11:36:42 -0500]
[INFO][2018-12-11 11:36:42 -0500]	-----------------------------------Phase 2: Rendering template(s)-----------------------------------
[INFO][2018-12-11 11:36:42 -0500]	Rendering yaml-error.yml ...
[INFO][2018-12-11 11:36:42 -0500]
[INFO][2018-12-11 11:36:42 -0500]	------------------------------------------Result: FAILURE-------------------------------------------
[FATAL][2018-12-11 11:36:42 -0500]	Invalid template: yaml-error.yml
[FATAL][2018-12-11 11:36:42 -0500]	Error message: Template is not valid YAML. (<rendered> yaml-error.yml): mapping values are not allowed in this context at line 9 column 21
```